tests/e2e/json_utils.go

```diff
@@ -0,0 +1,412 @@
+		err := json.Unmarshal(rawAction, &a)
+		if err == nil {
+			return a, nil
+		}
+	case "main.doublesignSlashAction":
+		var a doublesignSlashAction
+		err := json.Unmarshal(rawAction, &a)
+		if err == nil {
+			return a, nil
+		}
+	case "main.startRelayerAction":
+		var a startRelayerAction
+		err := json.Unmarshal(rawAction, &a)
+		if err == nil {
+			return a, nil
+		}
+	case "main.slashThrottleDequeue":
+		var a slashThrottleDequeue
+		err := json.Unmarshal(rawAction, &a)
+		if err == nil {
+			return a, nil
+		}
+	case "main.createIbcClientsAction":
+		var a createIbcClientsAction
+		err := json.Unmarshal(rawAction, &a)
+		if err == nil {
+			return a, nil
+		}
+	case "main.LegacyUpgradeProposalAction":
+		var a LegacyUpgradeProposalAction
+		err := json.Unmarshal(rawAction, &a)
+		if err == nil {
+			return a, nil
+		}
+	case "main.waitUntilBlockAction":
+		var a waitUntilBlockAction
+		err := json.Unmarshal(rawAction, &a)
+		if err == nil {
+			return a, nil
+		}
+	case "main.ChangeoverChainAction":
+		var a ChangeoverChainAction
+		err := json.Unmarshal(rawAction, &a)
+		if err == nil {
+			return a, nil
+		}
+	case "main.StartSovereignChainAction":
+		var a StartSovereignChainAction
+		err := json.Unmarshal(rawAction, &a)
+		if err == nil {
+			return a, nil
+		}
+	case "main.lightClientEquivocationAttackAction":
+		var a lightClientEquivocationAttackAction
+		err := json.Unmarshal(rawAction, &a)
+		if err == nil {
+			return a, nil
+		}
+	case "main.lightClientAmnesiaAttackAction":
+		var a lightClientAmnesiaAttackAction
+		err := json.Unmarshal(rawAction, &a)
+		if err == nil {
+			return a, nil
+		}
+	case "main.lightClientLunaticAttackAction":
+		var a lightClientLunaticAttackAction
+		err := json.Unmarshal(rawAction, &a)
+		if err == nil {
+			return a, nil
+		}
+	default:
+		return nil, fmt.Errorf("unknown action name: %s", actionTypeString)
+	}
+	return nil, err
+}
+
+// custom marshal and unmarshal functions for the chainstate that convert proposals to/from the auxiliary type with type info
+
+// transform the ChainState into a ChainStateWithProposalTypes by adding type info to the proposals
+func (c ChainState) MarshalJSON() ([]byte, error) {
+	type ProposalAndType struct {
+		RawProposal interface{}
+		Type        string
+	}
+
+	type ChainStateWithProposalTypes struct {
+		ValBalances                    *map[ValidatorID]uint
+		ValPowers                      *map[ValidatorID]uint
+		RepresentativePowers           *map[ValidatorID]uint
+		Params                         *[]Param
+		Rewards                        *Rewards
+		ConsumerChains                 *map[ChainID]bool
+		AssignedKeys                   *map[ValidatorID]string
+		ProviderKeys                   *map[ValidatorID]string
+		ConsumerChainQueueSizes        *map[ChainID]uint
+		GlobalSlashQueueSize           *uint
+		RegisteredConsumerRewardDenoms *[]string
+		Proposals                      *map[uint]ProposalAndType // the only thing changed from the real ChainState
```

Comment:
Exactly, I need to change the type of proposals. Happy about other suggestions, but I didn't find a good way

Reply:
@insumity the `ClientsFrozenHeights` chain state will be introduced by the `feat/ics-misbehaviour-handling` merge into `main`.

>Exactly, I need to change the type of proposals. 
@p-offtermatt I would check if field shadowing can solve this problem: https://stackoverflow.com/questions/56982842/how-to-selectively-marshal-json-for-a-struct.

Reply:
The problem with shadowing is that it will recursively try to marshal. Note that I am overriding the ChainState.Marshal/Unmarshal functions here. If I have a ChainState object as part of the ChainStateWithProposalTypes object, we have an endless loop. I think I could maybe use shadowing here, but I think that just shifts the complexity around to now getting the new ChainStateWithProposalType to marshal correctly.